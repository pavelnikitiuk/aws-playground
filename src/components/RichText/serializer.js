import React from 'react';
import { Elements } from 'prismic-reactjs';
import {
  Text, Heading, Link, List, ListIcon, ListItem,
} from '@chakra-ui/react';
import { SettingsIcon } from '@chakra-ui/icons';

export const serializer = {
  [Elements.heading1]: {
    Component: Heading,
    props: {
      as: 'h1',
      size: '4xl',
      mb: '5',
    },
  },
  [Elements.heading2]: {
    Component: Heading,
    props: {
      as: 'h2',
      size: '2xl',
      mb: '5',
    },
  },
  [Elements.heading3]: {
    Component: Heading,
    props: {
      as: 'h3',
      size: 'xl',
      mb: '2',
    },
  },
  [Elements.heading4]: {
    Component: Heading,
    props: {
      as: 'h4',
      size: 'l',
    },
  },
  [Elements.paragraph]: {
    Component: Text,
    props: {
      fontSize: 'xl',
    },
  },
  [Elements.hyperlink]: {
    render: (element, content, children, key) => (
      <Link
        color="teal.500"
        key={key}
        href={element.data.url}
        isExternal={!!element.data.target}
      >
        {content}
      </Link>
    ),
  },
  [Elements.list]: {
    Component: List,
  },
  [Elements.listItem]: {
    render: (element, content, children, key) => (
      <ListItem
        key={key}
        m={2}
      >
        <ListIcon as={SettingsIcon} color="green.500" />
        {children}
      </ListItem>
    ),
  },
  [Elements.em]: {
    Component: Text,
    props: {
      as: 'i',
    },
  },
};
