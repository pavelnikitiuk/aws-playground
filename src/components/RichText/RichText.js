import React from 'react';
import {
  shape, arrayOf, string, array,
} from 'prop-types';
import { RichText as PrismicRichText } from 'prismic-reactjs';
import { serializer as defaultSerializer } from './serializer';

export const createSerializer = (serializers) => (type, element, content, children, key) => {
  const serializer = serializers[type];

  if (!serializer) {
    return null;
  }

  const { Component, props = {}, render } = serializer;

  if (render) {
    return render(element, content, children, key);
  }

  return (
    <Component key={key} {...props}>
      {children}
    </Component>
  );
};

function RichText({ text }) {
  return (
    <PrismicRichText
      render={text}
      htmlSerializer={createSerializer(defaultSerializer)}
    />
  );
}

export const textPropTypes = arrayOf(
  shape({
    type: string,
    text: string,
    // eslint-disable-next-line react/forbid-prop-types
    spans: array,
  }),
);

RichText.propTypes = {
  text: textPropTypes.isRequired,
};

export default RichText;
