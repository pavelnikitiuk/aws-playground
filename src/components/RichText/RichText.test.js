import React from 'react';
import renderer from 'react-test-renderer';
import RichText from './index';

const MOCK_TEXT_DATA = [
  {
    type: 'paragraph',
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    spans: [
      {
        start: 0,
        end: 15,
        type: 'hyperlink',
        data: { link_type: 'Web', url: 'tel:+123456' },
      },
    ],
  },
  {
    type: 'paragraph',
    text: 'sed do eiusmod tempor incididunt ut labore et dolore magna aliqua',
    spans: [
      {
        start: 0,
        end: 24,
        type: 'hyperlink',
        data: { link_type: 'Web', url: '#', target: '_blank' },
      },
    ],
  },
  {
    type: 'heading1',
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    spans: [
      {
        start: 0,
        end: 8,
        type: 'hyperlink',
        data: { link_type: 'Web', url: '#', target: '_blank' },
      },
    ],
  },
  {
    type: 'heading3',
    text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
    spans: [
      {
        start: 0,
        end: 6,
        type: 'hyperlink',
        data: { link_type: 'Web', url: '#', target: '_blank' },
      },
    ],
  },
  { type: 'list-item', text: 'list-item;', spans: [] },
];

describe('Given the RichText component', () => {
  describe('when it is rendered', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = renderer.create(<RichText text={MOCK_TEXT_DATA} />);
    });

    it('should match the snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
