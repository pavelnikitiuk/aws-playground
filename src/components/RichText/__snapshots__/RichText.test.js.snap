// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the RichText component when it is rendered should match the snapshot 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-6dvxm2{font-size:xl;}",
      }
    }
    data-emotion="css 6dvxm2"
  />,
  <p
    className="chakra-text css-6dvxm2"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1856l8r{color:teal.500;}",
        }
      }
      data-emotion="css 1856l8r"
    />
    <a
      className="chakra-link css-1856l8r"
      href="tel:+123456"
    >
      Lorem ipsum dol
    </a>
    or sit amet, consectetur adipiscing elit
  </p>,
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-6dvxm2{font-size:xl;}",
      }
    }
    data-emotion="css 6dvxm2"
  />,
  <p
    className="chakra-text css-6dvxm2"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1856l8r{color:teal.500;}",
        }
      }
      data-emotion="css 1856l8r"
    />
    <a
      className="chakra-link css-1856l8r"
      href="#"
      rel="noopener noreferrer"
      target="_blank"
    >
      sed do eiusmod tempor in
    </a>
    cididunt ut labore et dolore magna aliqua
  </p>,
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-r7q7qr{margin-bottom:5px;}",
      }
    }
    data-emotion="css r7q7qr"
  />,
  <h1
    className="chakra-heading css-r7q7qr"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1856l8r{color:teal.500;}",
        }
      }
      data-emotion="css 1856l8r"
    />
    <a
      className="chakra-link css-1856l8r"
      href="#"
      rel="noopener noreferrer"
      target="_blank"
    >
      Lorem ip
    </a>
    sum dolor sit amet, consectetur adipiscing elit
  </h1>,
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-wlea3r{margin-bottom:2px;}",
      }
    }
    data-emotion="css wlea3r"
  />,
  <h3
    className="chakra-heading css-wlea3r"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-1856l8r{color:teal.500;}",
        }
      }
      data-emotion="css 1856l8r"
    />
    <a
      className="chakra-link css-1856l8r"
      href="#"
      rel="noopener noreferrer"
      target="_blank"
    >
      Lorem 
    </a>
    ipsum dolor sit amet, consectetur adipiscing elit
  </h3>,
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ".css-155za0w{list-style-type:none;}",
      }
    }
    data-emotion="css 155za0w"
  />,
  <ul
    className="css-155za0w"
    role="list"
  >
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ".css-7gd3gi{margin:2px;}",
        }
      }
      data-emotion="css 7gd3gi"
    />
    <li
      className="css-7gd3gi"
    >
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-qld8i7{width:1em;height:1em;display:inline-block;line-height:1em;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;color:green.500;}",
          }
        }
        data-emotion="css qld8i7"
      />
      <style
        dangerouslySetInnerHTML={
          Object {
            "__html": ".css-16alosh{width:1em;height:1em;display:inline-block;line-height:1em;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;color:currentColor;vertical-align:middle;width:1em;height:1em;display:inline-block;line-height:1em;-webkit-flex-shrink:0;-ms-flex-negative:0;flex-shrink:0;color:green.500;}",
          }
        }
        data-emotion="css 16alosh"
      />
      <svg
        className="chakra-icon chakra-icon css-16alosh"
        focusable={false}
        role="presentation"
        viewBox="0 0 14 14"
      >
        <path
          d="M14,7.77 L14,6.17 L12.06,5.53 L11.61,4.44 L12.49,2.6 L11.36,1.47 L9.55,2.38 L8.46,1.93 L7.77,0.01 L6.17,0.01 L5.54,1.95 L4.43,2.4 L2.59,1.52 L1.46,2.65 L2.37,4.46 L1.92,5.55 L0,6.23 L0,7.82 L1.94,8.46 L2.39,9.55 L1.51,11.39 L2.64,12.52 L4.45,11.61 L5.54,12.06 L6.23,13.98 L7.82,13.98 L8.45,12.04 L9.56,11.59 L11.4,12.47 L12.53,11.34 L11.61,9.53 L12.08,8.44 L14,7.75 L14,7.77 Z M7,10 C5.34,10 4,8.66 4,7 C4,5.34 5.34,4 7,4 C8.66,4 10,5.34 10,7 C10,8.66 8.66,10 7,10 Z"
          fill="currentColor"
        />
      </svg>
      list-item;
    </li>
  </ul>,
]
`;
